book/src/libraries/api_lib_guide.md

```diff
@@ -0,0 +1,319 @@
+- It has no backing struct
+- values defined inside it will only persist for the duration of the function call
+
+Example:
+
+```iecst
+FUNCTION myFunc : DINT
+VAR_INPUT
+  x  : DINT;
+END_VAR
+END_FUNCTION
+```
+```c
+int32_t myFunc(int32_t);
+```
+
+In contrast, a `FUNCTION_BLOCK` is backed by a struct and is globally accessible by a defined instance.
+To declare a `FUNCTION_BLOCK`, a backing struct has to be declared and passed as a reference to the function block implementation.
+
+```iecst
+FUNCTION_BLOCK myFb
+VAR_INPUT
+  x  : DINT;
+END_VAR
+END_FUNCTION_BLOCK
+```
+
+```c
+typedef struct {
+  int32_t x;
+} myFunctStr;
+
+void myFb(myFunctStr*);
+```
+
+
+
+#### 2.2.1 Parameters
+
+`FUNCTION` and `FUNCTION_BLOCK` may define input parameters. These are passed using the `VAR_INPUT` or `VAR_IN_OUT` blocks.
+The difference between the two blocks is how the values are passed.
+A `VAR_INPUT` variable is passed _by value_, while a `VAR_IN_OUT` variable is passed _by reference_.
+In general, it is recommended to use a `VAR_IN_OUT` for data that needs to be both read and written, while `VAR_INPUT` should be reserved to read only values.
+
+> **_NOTE:_** In `FUNCTION`s complex datatypes are handled as pointers.
+
+Examples:
+
+`FUNCTION`:
+
+```iecst
+FUNCTION myFunc : DINT
+VAR_INPUT
+  myInt  : DINT;
+  myString : STRING[255];
+END_VAR
+
+VAR_INPUT
+  myRefStr : STRING;
+END_VAR
+
+VAR_IN_OUT
+  myInOutInt : DINT;
+END_VAR
+END_FUNCTION
+```
+
+```c
+int32_t myFunc(int32_t myInt, char* myString, char* myRefStr, int32_t* myInOutInt);
+```
+
+`FUNCTION_BLOCK`:
+
+
+```iecst
+FUNCTION_BLOCK myFb
+VAR_INPUT
+  myInt  : DINT;
+  myString : STRING[255];
+END_VAR
+
+VAR_IN_OUT
+  myInOutInt : DINT;
+END_VAR
+END_FUNCTION_BLOCK
+```
+
+```c
+
+typedef struct {
+  int32_t myInt;
+  char myString[256];
+  int32_t* myInOutInt;
+} myFbStruct
+
+void myFb(myFbStruct* myFbInstance);
+```
+
+
+#### 2.2.2 Private members
+
+A `FUNCTION_BLOCK` often requires local (private) members to hold data across executions. These members have to be declared in the struct.
+As a side effect, these variables are visible to the users.
+
+For example:
+
+```
+FUNCTION_BLOCK Count
+VAR
+  current : DINT;
+END_VAR
+END_FUNCTION_BLOCK
+```
+
+```c
+
+typedef struct {
+  int32_t current;
+} CountStruct;
+
+void Count(CountStruct* countInst) {
+  countInst->current = countInst->current + 1;
+}
+```
+
+
+
+#### 2.2.3 Return values
+
+A `FUNCTION` defines a return value in the signature, while a `FUNCTION_BLOCK` relies on `VAR_OUTPUT` definitions.
+
+The return type for a function can also include complex datatypes, such as strings, arrays and structs.
+Internally, complex return types are treated as reference parameters.
+
+Example:
+
+```iecst
+FUNCTION myFunc : STRING
+  VAR_INPUT
+    x : DINT;
+  END_VAR
+  VAR_IN_OUT
+    y : DINT;
+  END_VAR
+END_FUNCTION
+```
+
+The C interface would look like:
+
+```c
+void myFunc(char* out, int32_t x, int32_t* y);
```

Comment:
Maybe it should be made clearer that this is the return value and why the C function is void and has not the return type String

Reply:
To add to this, would it make sense to mention how the equivalent C code looks like for primitive datatypes followed by complex ones such as the given string?